cggmp21/src/signing.rs

```diff
@@ -715,13 +715,12 @@ where
         )
         .map_err(|e| Bug::PiEnc(BugSource::psi0, e))?;
 
-        tracer.send_msg();
         outgoings
-            .send(Outgoing::p2p(j, Msg::Round1b(MsgRound1b { psi0 })))
+            .feed(Outgoing::p2p(j, Msg::Round1b(MsgRound1b { psi0 })))
             .await
             .map_err(IoError::send_message)?;
-        tracer.msg_sent();
     }
+    outgoings.flush().await.map_err(IoError::send_message)?;
```

Comment:
You should keep `tracer.send_msg()`/`tracer.msg_sent()` so the time we spend on sending messages gets counted

Reply:
We can rename `send_msg`/`msg_sent` function names so the naming doesn't imply that a message has to be sent between these calls

Reply:
In fact, the only implementation of the trace trait does not assume that a single message actually gets sent between these calls. The time between these calls just get accounted as spent on sending messages, no matter how many were sent.

Reply:
> In fact, the only implementation of the trace trait we have does not assume that a single message actually gets sent between these calls

Ugh, ok, that's a weird metric then. But I'll put the traces back in this case

Reply:
Measuring how much time we spend on sending messages in a weird metric to you?

Reply:
That's a good metric. But we're not measuring it here, what we'll be measuring is time spent adding messages to a queue. I don't find it very useful, especially with a misleading name. Maybe it makes sense to add some trace to flushing though.

Reply:
We will be measuring time spent on adding to queue + time spent on flushing items from the queue. Two of them makes the time we spend on sending messages.

Reply:
> Two of them makes

I expect that feeding messages makes them be processed in the background, io_uring/python asyncio style. So the two of the measurements is not greater than the time spent on sending messages

Reply:
> I expect that feeding messages makes them be processed in the background

`feed` does the same thing as `send`, it just doesn't do `flush` in the end

Reply:
There's no magical background processors for sinks. Either sink itself supports that or not. If it doesn't, then messages are sent within protocol.

Sink can indeed send messages to other task to do sending, but then caller should not expect `Tracer` to be able to measure time we spend on sending messages, as protocol does not get to observe this.

Reply:
> I expect that feeding messages makes them be processed in the background, io_uring/python asyncio style

You shouldn't be expecting that, it's not python, nor was it inspired by io_uring :smiley:

Reply:
> There's no magical background processors for sinks. Either sink itself supports that or not. If it doesn't, then messages are sent within protocol

Exactly. And here we're writing code for any sink, so a sink that does background processing will show up weird in the tracer. 

> caller should not expect Tracer to be able to measure time we spend on sending messages

That's exactly what I'm telling you: depending on the sink implementation, these sending timings are completely wrong